Source/SwiftLintFramework/Rules/IdentifierNameRule.swift

```diff
@@ -122,15 +121,11 @@ private extension String {
         guard firstCharacter.isUppercase() else {
             return false
         }
+
```

Comment:
Maybe this code section needs a general overhaul - at the moment it also looks like you can dodge it by starting an identifier with two uppercase letters, like `let UIcolor`. Maybe this could be an alternative:

```
extension String {
    var isViolatingCase: Bool {
        guard let firstUnicodeScalar = unicodeScalars.first else { return false }

        if CharacterSet.lowercaseLetters.contains(firstUnicodeScalar) {
            return false
        } else if CharacterSet.uppercaseLetters.contains(firstUnicodeScalar) {
            guard characters.count > 1 else { return true } // disallow single uppercases letters
            return unicodeScalars.contains(where: CharacterSet.lowercaseLetters.contains)
        } else {
            // Only allow if no letters at all (operator-ish)
            guard characters.count > 1 else { return false } // optimization
            return unicodeScalars.contains(where: CharacterSet.letters.contains)
        }
    }
}
```

```
// okay
"".isViolatingCase
"f".isViolatingCase
"f1".isViolatingCase
"_".isViolatingCase
"__".isViolatingCase
"FOO".isViolatingCase
"F1".isViolatingCase
"SOME_SNAKE_CASE".isViolatingCase

// violating
"F".isViolatingCase
"Foo".isViolatingCase
"_f".isViolatingCase
"_F".isViolatingCase
```

Reply:
I think SwiftLint allowed `_f` before though. So in that case we could adjust the last `else` branch to skip non-letters at the beginning and then do a recursive `isViolatingCase` on the rest.


- `_fooBar` -> `fooBar` -> okay
- `_FooBar` -> `FooBar` -> violation
- `___FOO_BAR` -> `FOO_BAR` -> okay

Reply:
```
extension String {
    var isViolatingCase: Bool {
        guard let firstUnicodeScalar = unicodeScalars.first else { return false }

        if CharacterSet.lowercaseLetters.contains(firstUnicodeScalar) {
            return false
        } else if CharacterSet.uppercaseLetters.contains(firstUnicodeScalar) {
            guard characters.count > 1 else { return true } // disallow single uppercases letters
            return unicodeScalars.contains(where: CharacterSet.lowercaseLetters.contains)
        } else {
            guard
                characters.count > 1,
                case let casedLetters = CharacterSet.lowercaseLetters.union(.uppercaseLetters),
                let casedLetterIndex = unicodeScalars.index(where: casedLetters.contains) else {
                    // Allow if no cased letters at all (operator-ish)
                    return false
            }

            return substring(from: casedLetterIndex).isViolatingCase
        }
    }
}
```

```
// okay
"".isViolatingCase
"f".isViolatingCase
"f1".isViolatingCase
"_".isViolatingCase
"__".isViolatingCase
"FOO".isViolatingCase
"F1".isViolatingCase
"SOME_SNAKE_CASE".isViolatingCase
"_f".isViolatingCase
"§".isViolatingCase
"•".isViolatingCase
"_fooBar".isViolatingCase
"___FOO_BAR".isViolatingCase

// violating
"F".isViolatingCase
"Foo".isViolatingCase
"_F".isViolatingCase
"_FooBar".isViolatingCase
```

Reply:
I'm not super happy with `F1` being valid, but to fix that we would have to adjust the "uppercase is okay if at least two characters and all upper case" to a more strict "uppercase is okay if at least 2 *cased* letters". In that case `FO1` would be okay but `F1` not. Don't know if worth the hassle though.

Reply:
This would be the variant implementing this check:

```
extension String {
    var isViolatingCase: Bool {
        guard let firstUnicodeScalar = unicodeScalars.first else { return false }

        if CharacterSet.lowercaseLetters.contains(firstUnicodeScalar) {
            return false
        } else if CharacterSet.uppercaseLetters.contains(firstUnicodeScalar) {
            guard
                characters.count > 1, // optimization
                case let rest = unicodeScalars.suffix(from: index(after: startIndex)),
                rest.contains(where: CharacterSet.uppercaseLetters.contains) else {
                    // Disallow if only a single uppercase letter
                    return true
            }

            return unicodeScalars.contains(where: CharacterSet.lowercaseLetters.contains)
        } else {
            guard
                characters.count > 1, // optimization
                case let casedLetters = CharacterSet.lowercaseLetters.union(.uppercaseLetters),
                let casedLetterIndex = unicodeScalars.index(where: casedLetters.contains) else {
                    // Allow if no cased letters at all (operator-ish)
                    return false
            }

            return substring(from: casedLetterIndex).isViolatingCase
        }
    }
}
```

```
// okay
"".isViolatingCase
"f".isViolatingCase
"f1".isViolatingCase
"_".isViolatingCase
"__".isViolatingCase
"FOO".isViolatingCase
"F_O".isViolatingCase
"SOME_SNAKE_CASE".isViolatingCase
"_f".isViolatingCase
"§".isViolatingCase
"•".isViolatingCase
"_fooBar".isViolatingCase
"___FOO_BAR".isViolatingCase

// violating
"F".isViolatingCase
"F1".isViolatingCase
"F__".isViolatingCase
"Foo".isViolatingCase
"_F".isViolatingCase
"_FooBar".isViolatingCase
```

Doesn't look too bad, should be fast, and seems quite reasonable to me - what do you think?

Reply:
@SDGGiesbrecht Oh, good point! What I wanted to avoid is another hard coded list like `_` (since this PR tries to remove one). How do you feel about this one (also incorporating an extension on `Character`)?

```swift
extension Character {
    enum Casing {
        case lower, upper
    }

    var casing: Casing? {
        let unicodeScalars = String(self).unicodeScalars

        if unicodeScalars.contains(where: CharacterSet.lowercaseLetters.contains) {
            return .lower
        } else if unicodeScalars.contains(where: CharacterSet.uppercaseLetters.contains) {
            return .upper
        } else {
            return nil
        }
    }
}

extension String {
    var isViolatingCase: Bool {
        guard let firstCharacter = characters.first else { return false }

        switch firstCharacter.casing {
        case .lower?:
            return false
        case .upper?:
            guard characters.count > 1 else { return true } // optimization
            var numberOfUppercaseLetters = 1

            for character in characters.suffix(from: index(after: startIndex)) {
                switch character.casing {
                case .lower?: return true
                case .upper?: numberOfUppercaseLetters += 1
                case nil: break
                }
            }

            return numberOfUppercaseLetters < 2
        case nil:
            guard
                characters.count > 1, // optimization
                let firstCasedLetter = characters.index(where: { $0.casing != nil }) else {
                    // Allow if no cased letters at all (operator-ish)
                    return false
            }

            return substring(from: firstCasedLetter).isViolatingCase
        }
    }
}
```

Tests:

```swift
// okay
"".isViolatingCase
"f".isViolatingCase
"f1".isViolatingCase
"_".isViolatingCase
"__".isViolatingCase
"FOO".isViolatingCase
"F_O".isViolatingCase
"SOME_SNAKE_CASE".isViolatingCase
"_f".isViolatingCase
"§".isViolatingCase
"•".isViolatingCase
"_fooBar".isViolatingCase
"___FOO_BAR".isViolatingCase
"é".isViolatingCase
"ÉÉ".isViolatingCase
"ß".isViolatingCase
"Ç̌E".isViolatingCase

// violating
"F".isViolatingCase
"F1".isViolatingCase
"F__".isViolatingCase
"Foo".isViolatingCase
"_F".isViolatingCase
"_FooBar".isViolatingCase
"É".isViolatingCase
"Ç̌".isViolatingCase
```

Reply:
Much more concise version:

```
extension String {
    var isViolatingCase: Bool {
        var numberOfUppercaseLetters = 0

        for character in characters {
            switch character.casing {
            case .lower?:
                return numberOfUppercaseLetters > 0
            case .upper?:
                numberOfUppercaseLetters += 1
            case nil:
                break
            }
        }

        return numberOfUppercaseLetters == 1
    }
}
```

Reply:
Valiant attempt, @sirlantis, but Unicode is more complicated than that. The `Casing` enumeration is insufficient: a `Character` may actually contain `lower`, `upper`, **both** or neither (your `nil`), so `var ǅblahblahblah = ...` winds up being allowed as lowercase. There are a few other quirks leaking through as well, but...

The abundance comments have made the reviews hard to read illegible, so I will continue the conversation in a pull request to @marcelofabri’s branch instead. I also think his original code is very concise; it only needs minor tweaks, not a complete re‐write.